src/test/java/bsh/InterpreterTest.java

```diff
@@ -52,4 +55,61 @@ public void check_system_object() throws Exception {
         assertEquals(Boolean.FALSE, TestUtil.eval("return bsh.system.shutdownOnExit;"));
     }
 
+    @Test
+    public void interrupt_on_tokenizing() throws Exception {
+        final StringReader in = new StringReader("print(\"\n");
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        Interpreter bsh = new Interpreter(in, new PrintStream(baos), System.err, false);
+
+        Thread t = new Thread(bsh);
+        t.start();
+
+        Thread.sleep(250); // give Interpreter eom time to set itself up
+
+        //
+        // now the interpreter should be on a blocking read; we interrupt it,
+        // give it some time to make the magic and make sure it tells us about it
+        //
+        bsh.interrupt(); Thread.sleep(250);
+
+        assertTrue(baos.toString().contains("... interrupted ..."));
+        assertTrue(!baos.toString().contains("Parser Error:"));
+
+        //
+        // we should also be able to issue new statements, but this would require
+        // some more refactoring, let's defer it to the next iteration
+        //
+    }
+
+    @Test
+    public void interrupt_on_parsing() throws Exception {
+        final StringReader in = new StringReader("class A {\n"); // generate a ParseException
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final ByteArrayOutputStream err = new ByteArrayOutputStream();
+
+        Interpreter bsh = new Interpreter(in, new PrintStream(out), new PrintStream(err), false);
+
+        //
+        // A ParseException after interrupt() is most likely due to the
+        // interruption and the error should be muted
+        //
+        bsh.interrupt(); // to make it easy let's make sure we start interrupted
+
+        Thread t = new Thread(bsh);
+        t.start();
+
+        Thread.sleep(250); // give Interpreter eom time to set itself up
```

Comment:
This is how I got iTerm to drop the prompt, by trying to slow the main thread. =(

Reply:
That I think is ok for th tests only. What I have done in BshConsoleInterpeter is: 

in startConsole():

synchronized (this) {
            try {
                wait(2000);
            } catch (InterruptedException x) {
                // nop
            }
        }

in prompt():
synchronized (this) {
            notifyAll();
        }

This can probably only achieved outside Interpreter...